### ISSUE TITLE
Hello, distilling clip on yfcc dataset of 1 million image-text pairs, training 5 rounds of accuracy only 0.0011. could you please give me some advice? thanks!!

### ISSUE THREAD

**Issue by leo23ui** [microsoft/Cream#249]
Hello, distilling clip on yfcc dataset of 1 million image-text pairs, training 5 rounds of accuracy only 0.0011, and the gpu memory is only  15728MiB , the volatile gpu utilize is 0%, is this normal? could you please give me some advice? thanks!! I suspect there is something wrong with the way I read tar files. My sh file looks like this

export NNODES=1
export GPUS_PER_NODE=1
export WANDB__SERVICE_WAIT=60  

DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES"
torchrun $DISTRIBUTED_ARGS src/training/main.py \
 --save-frequency 1 \
 --report-to wandb \
 --train-data /home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/ \
 --dataset-type webdataset \
 --imagenet-val ./ImageNet \
 --warmup 2000 \
 --batch-size 1024 \
 --epochs 8 \
 --workers 1 \
 --model TinyCLIP-ViT-39M-16-Text-19M \
 --name exp_name \
 --seed 0 \
 --local-loss \
 --grad-checkpointing \
 --output ./outputs/TinyCLIP-ViT-39M-16-Text-19M \
 --lr 0.0001 \
 --gather-with-grad \
 --pretrained-image-file ViT-B-16@openai \
 --pretrained-text-file ViT-B-16@openai \
 --distillation-teacher ViT-B-32@laion2b_e16 \
 --norm_gradient_clip 5 \
 --train-num-samples  1000000 \
 --logit-scale 50

**Comment by wkcn** [comment]
Hi @leo23ui , it is abnormal. Please check the dataset and the training loss.

**Comment by leo23ui** [comment]
> Hi [@leo23ui](https://github.com/leo23ui) , it is abnormal. Please check the dataset and the training loss.

thanks for your reply!!! the training loss from 4.3 to 3.9 in 8 epoches
Is the dataset format for the original code(tinyclip) multiple tar files, or are all yfcc15m dataset image-text pairs all in one tar? Chatgpt told me wds.SimpleShardList can load single tar file or a diretory's tar files. I can't compress all the data sets here, could you give me some suggestions? Thanks

In sh, I use "--train-data /home/gg/gg/MQBench-main/test/model/e/split_1"which is a directory contains many tar files. I use the following code to output input_shards to help me debug. The input_shards is a list datatype with a length of 980 and each element is strdatatype represent tar file path, as shown in the following lines {} . According to chatgpt, wds.SimpleShardList(input_shards) can load list which contains many str(the paths  to many tar files) files smoothly, but I don't know why the above gpu has the problems of 14531mb gpu memory, 0% gpu volatile-util, low accuracy and great loss as shown in the above comment. could you give me some suggestions? Thanks   

///
def get_tar_files_from_folder(folder):  
    return sorted(glob.glob(os.path.join(folder, "*.tar")))       
input_shards = get_tar_files_from_folder(args.train_data)      
    for shard in input_shards:
        print(shard)  

pipeline = [wds.SimpleShardList(input_shards)]
///
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_969.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_97.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_970.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_971.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_972.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_973.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_974.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_975.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_976.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_977.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_978.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_979.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_98.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_980.tar'}
{'url': '/home/gg/gg/MQBench-main/test/model/e/split_1/split_1tar/batch_99.tar'}

**Comment by leo23ui** [comment]
I suspect that it may be because the image loading failure led to the mismatch of the image and text, but the select function can delete the mismatched data, and it will not affect the gpu, there should be other reasons, could you please give me some suggestions? Thanks!!!


2024-11-18,05:49:26 | WARNING | Handling webdataset error (OSError('image file is truncated (76 bytes not processed)')). Ignoring.                           [675/1306]

**Comment by wkcn** [comment]
@leo23ui I am not familiar with the webdataset.
You can try to visualize the image and the text in a pair provided by the dataloader.
```python
for batch in dataloader:
    images, texts = batch
    # images is a tensor with the shape of (B, C, H, W) after image transform
    # texts is a tensor with the shape of (B, #tokens such as 77)
    # You can de-transform the images and de-tokenize the texts
```

**Comment by leo23ui** [comment]
In addition,  i set the --train-num-samples  5000000, but 



> [@leo23ui](https://github.com/leo23ui) I am not familiar with the webdataset. You can try to visualize the image and the text in a pair provided by the dataloader.
> 
> for batch in dataloader:
>     images, texts = batch
>     # images is a tensor with the shape of (B, C, H, W) after image transform
>     # texts is a tensor with the shape of (B, #tokens such as 77)
>     # You can de-transform the images and de-tokenize the texts

I used the following code to verify whether i read the tar files successfully， and i find it outputs many filename from different tar files, i think the read of dataset is correct, i think there is some other reasons that causes low accracy and low gpu memory. could you please give me some advice, thanks!!

def check_filenames(sample):

    key = sample.get("__key__", "")
    image_key = f"{key}.jpg" 
    text_key = f"{key}.txt" 


    print(f"Key: {key}")
    print(f"Image Filename: {image_key}")
    print(f"Text Filename: {text_key}")

    if image_key != text_key:
        print(f": image {image_key}, text {text_key}")
    else:
        print("same")

    return sample

 pipeline.extend([wds.map(check_filenames) ])

------------------------------------------------------

the output of check_filenames is as follows:

Key: 5669596963_a6360f191a
Image Filename: 5669596963_a6360f191a.jpg
Text Filename: 5669596963_a6360f191a.txt
: image 5669596963_a6360f191a.jpg, text 5669596963_a6360f191a.txt

Key: 13663699084_0c21243bbc
Image Filename: 13663699084_0c21243bbc.jpg
Text Filename: 13663699084_0c21243bbc.txt
: image 13663699084_0c21243bbc.jpg, text 13663699084_0c21243bbc.txt

Key: 136361811_9212f23522
Image Filename: 136361811_9212f23522.jpg
Text Filename: 136361811_9212f23522.txt
: image 136361811_9212f23522.jpg, text 136361811_9212f23522.txt
